torch/csrc/jit/autodiff.cpp

```diff
@@ -0,0 +1,100 @@
+  const auto build_sym_grad = [node](const std::vector<SymbolicVariable>& grads) -> std::vector<SymbolicVariable> {
+    auto inputs = node->inputs();
+    switch(node->kind()) {
+      case kadd:
+        return {grads[0], grads[0]};
+      case ksub:
+        return {grads[0], -grads[0]};
+      case kmul:
+        return {grads[0] * inputs[1], grads[0] * inputs[0]};
+    }
+    throw std::runtime_error(std::string("don't support differentiation of `") +
+                            node->kind().toString() + "`");
+  };
+  auto sym_grads = build_sym_grad(fmap<SymbolicVariable>(grad_values));
+  return fmap(sym_grads, [](const SymbolicVariable &v) { return v.value(); });
+}
+
+void differentiate(std::shared_ptr<Graph>& graph) {
+  JIT_ASSERT(graph->stage() == 0);
+  graph->advanceStage();
+
+  std::unordered_map<Value*, Value*> grad_map; // x -> dx mapping
+  const auto get_grad = [&](Value* v) { return grad_map[v]; };
+  for (auto output : graph->outputs())
+    grad_map[output] = graph->addInput()->setType(output->typeOption());
+
+  for (auto it = graph->rbegin(), end = graph->rend(); it != end; ++it) {
+    Node *node = *it;
+    auto inputs = node->inputs();
+    value_list grad_inputs = gradientForNode(node, fmap(node->outputs(), get_grad));
+    JIT_ASSERT(grad_inputs.size() == node->inputs().size());
+    for (std::size_t i = 0, num_inputs = grad_inputs.size(); i < num_inputs; ++i) {
+      if (Value * prev_grad = grad_map[inputs[i]]) {
+        Node *new_grad_node = graph->create(kadd, {prev_grad, grad_inputs[i]})
+                                   ->t_(kalpha, at::Scalar(1).toTensor());
+        new_grad_node->insertAfter(grad_inputs[i]->node());
+        Value *new_grad = new_grad_node->output();
+        new_grad->setType(prev_grad->typeOption());
+        grad_map[inputs[i]] = new_grad;
+      } else {
+        grad_map[inputs[i]] = grad_inputs[i];
+      }
+    }
+  }
+
+  for (auto input : graph->inputs()) {
+    if (input->stage() > 0) break;
+    graph->registerOutput(grad_map.at(input));
+  }
+}
+
+// TODO: return metadata that allows to associate inputs with values form graph
+std::pair<std::shared_ptr<Graph>, value_list> backwardLambdaLift(std::shared_ptr<Graph>& graph) {
+  JIT_ASSERT(graph->stage() == 1);
+
+  std::unordered_set<Value*> backward_inputs_set;
+  for (Node *node : *graph) {
+    if (node->stage() == 0) continue;
+    for (Value *input : node->inputs()) {
+      // XXX: this assumes that the single Param node in the graph (that holds inputs) has stage 0
+      if (input->node()->stage() == 0)
+        backward_inputs_set.insert(input);
+    }
+  }
+  value_list backward_inputs(backward_inputs_set.begin(), backward_inputs_set.end());
+  // We want them nicely sorted with inputs being first, and then following
+  // the topological ordering (at least approximately, we could do a better job here).
+  std::sort(backward_inputs.begin(), backward_inputs.end(),
```

Comment:
Let's do this in true topological order.  Having a confusing order for the captures will make it hard to debug.

Reply:
I'm a bit worried that we'll keep reconstructing these topo_idx maps over and over (they appear in 3 places now I think). On the other hand maintaining an ordering on a changing graph is annoying, but possible if only we had a simple balanced tree implementation. I don't think it can be implemented using `std`, so we'd need an extra dependency or to roll our own